### ISSUE TITLE
Handling unwanted or unsupported attestation formats

### ISSUE THREAD

**Issue by madwizard-thomas** [w3c/webauthn#1485]
It is not quite clear to me how a Relying Party should handle an attestation format it did not expect or requested.

The `attestation` field in `PublicKeyCredentialCreationOptions` can be used to specify whether or not attestation information is desired. Even though the Create method in $5.1.3  specifies that "none" should be handled by returning a None attestation statement, the enum `AttestationConveyancePreference` seems to hint that this value is more of a preference than a hard requirement. So would it be possible for a client to return a real attestation statement anyway?
Firefox on Windows 10 (unlike Chrome or Edge) seems to always include attestation statements even if the `attestation` field is set to `none`.
`$7.1 Registering a new credential` specifies to validate the statement as provided in the response (step 18/19) but makes no mention of whether attestation was requested or what the `attestation` field was set to.

Another issue is that the attestation statement format list in the spec is not exhaustive, new statement formats may arise in the future. A current example is Apple's new attestation format. RP's likely will not have implemented the new format yet, in this case Apple has not even registered the format yet even though iOS 14 will be released today.

`$7.1 Registering a new credential` does not mention what should be done in case the statement format is not supported by the RP. This could be treated as a fatal error but if the attestation can be sent without asking this would fail the ceremony unnecessarily. Also in a situation where 'best effort' attestation is applicable (allow optional attestation to determine security levels or for use in a risk engine) the RP may come across unknown formats but still want to use the credential.

So to summarize:

1. When `attestation` is set to `"none"`, is it allowed to return an attestation format other than None?
2. If 1 is allowed, how should RP's handle this situation? Should they validate the attestation statement or ignore it / treat it as None?
3. How should RP deal with unsupported statement formats? Is this RP policy, a fatal error, treat it as None?

**Comment by jcjones** [comment]
> Firefox on Windows 10 (unlike Chrome or Edge) seems to always include attestation statements even if the attestation field is set to none.

I've opened [Bug 1665466](https://bugzilla.mozilla.org/show_bug.cgi?id=1665466) at Firefox for the Firefox bug - sorry that it's in the way!

**Comment by emlun** [comment]
1. Authenticators will, in general, always return an attestation statement if they are capable, but the browser can filter it depending on the `attestation` argument. Technically, I think the browser must respect the `attestation` argument to be compliant, but the response is still well-formed if it doesn't. So this is the classic dilemma between rejecting any deviation from the spec in the interest of compliance, versus tolerating some discrepancies in the interest of compatibility. I don't think there is a best answer to that one.
2. There is no security difference between receiving a `"none"` attestation, vs. receiving and ignoring a full attestation.
3. If the RP requires a trusted attestation (most likely for financial/government institutions subject to legal regulations, or for enterprises' internal systems that know what authenticators they want to allow), then by definition it has to reject any attestation format it doesn't support. If the RP does not require attestation, it can just ignore it or maybe save it for later.

**Comment by akshayku** [comment]
@madwizard-thomas , Can you also specify Windows 10 version?

**Comment by Kieun** [comment]
> There is no security difference between receiving a "none" attestation, vs. receiving and ignoring a full attestation.

@emlun regarding this, the security might be equal between two. But, the core difference is whether we can get the authenticator model identifier (aaguid) or not. At least with this information, RP might refer the details of the authenticator from metadata.

**Comment by madwizard-thomas** [comment]
Thanks all for the clarification. For my own library I think accepting unsupported attestations as an 'unknown' type with the verification resulting in a "none" attestation type is a valid approach then, with an option to enable strict compliance to the supported attestation types.

One of the reason's I use "best-effort" attestation is to determine the authenticator's device name and aid the user in automatically naming the registered key something like "Yubikey 5".

> I've opened [Bug 1665466](https://bugzilla.mozilla.org/show_bug.cgi?id=1665466) at Firefox for the Firefox bug - sorry that it's in the way!

Thanks, I suspected this was a bug but wasn't sure.

> @madwizard-thomas , Can you also specify Windows 10 version?

It's windows 10 Pro version 2004.

**Comment by emlun** [comment]
In case it helps, [Yubico's Java library](https://github.com/Yubico/java-webauthn-server) will always verify the attestation signature if present, and fail if the signature is invalid. However, by default it accepts any valid attestation, including `"none"`, but signals in the result that the attestation is valid but not trusted. If the user configures the library with a trust store _and_ enables the "require trusted attestation" setting, only then will the library reject any attestation that cannot be linked to some root certificate in the trust store.

**Comment by nicksteele** [comment]
To summarize thoughts discussed on the Working Group's call on this topic, if an RP receives an unwanted or unsupported attestation format, it should be handled at the discretion of the RP to be either ignored or handled in some other way. The specification or the client device should be explicit in saying whether unwarranted attestation data should be ignored or discarded.
